### PR TITLE
[RFC] Add support for Semi-Supervised active learning.

### PR DESCRIPTION
**Pull request by Dref360**
## Summary:

This RFC proposes support for semi-supervised learning in BaaL.
Papers such as [Gao et al.](https://arxiv.org/abs/1910.07153) propose to use semi-supervised learning with active learning to get better results. It is currently quite hard for users to use SS with BaaL.

### CODE REVIEW

File: rfcs/01-semisupervised.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    def train_on_pool_batch(self, data, target, optimizer, cuda=False):
+        if cuda:
+            data, target = to_cuda(data), to_cuda(target)
+        optimizer.zero_grad()
+        output = self.model(data)
+        loss = self.pool_criterion(data, output, target)
+        loss.backward()
+        optimizer.step()
+        self._update_metrics(output, target, loss, filter='train', is_labelled=False)
```
Comment (IssamLaradji) [inline rfcs/01-semisupervised.md]: looks solid, but I am slightly missing the point.  Will the following wrapper achieve the same thing? the user can then check if batch['is_labeled'] is true or not to decide whether to use the unsupervised or the supervised loss.
```
class DatasetWrapper(Dataset):
    def __init__(self, dataset, is_labeled):
        self.dataset = dataset
        self.is_labeled = is_labeled
        
    def __getitem__(self, index):
        data, target = self.dataset[index]


        return {"images":data, 
                'labels':target, 
                'is_labeled':self.is_labeled[index]}
```

Comment (Dref360) [reply rfcs/01-semisupervised.md]: Yeah under the hood it would probably be like that. Like ActiveLearningDataset.iter_alternate would return a Dataset.

**Comment by Dref360** [comment]
The RFC is accepted and PRs are welcome. Internally, @rafapi will work on it.
